sinks/signalfx/signalfx.go

```diff
@@ -42,28 +42,65 @@ func (c *collection) addPoint(key string, point *datapoint.Datapoint) {
 	c.points = append(c.points, point)
 }
 
-func (c *collection) submit(ctx context.Context, cl *trace.Client) error {
+func submitDatapoints(ctx context.Context, cl *trace.Client, client dpsink.Sink, points []*datapoint.Datapoint, errs chan<- error) {
+	span, ctx := trace.StartSpanFromContext(ctx, "")
+	span.SetTag("datapoint_count", len(points))
+	defer span.ClientFinish(cl)
+
+	err := client.AddDatapoints(ctx, points)
+	if err != nil {
+		span.Error(err)
+		span.Add(ssf.Count("flush.error_total", 1, map[string]string{"cause": "io", "sink": "signalfx"}))
+		errs <- err
+	}
+	errs <- nil
+}
+
+func (c *collection) submit(ctx context.Context, cl *trace.Client, maxPerFlush int) error {
 	wg := &sync.WaitGroup{}
 	errorCh := make(chan error, len(c.pointsByKey)+1)
 
-	submitOne := func(client dpsink.Sink, points []*datapoint.Datapoint) {
+	submitBatch := func(client dpsink.Sink, points []*datapoint.Datapoint) {
 		span, childCtx := trace.StartSpanFromContext(ctx, "")
 		span.SetTag("datapoint_count", len(points))
 		defer span.ClientFinish(cl)
 		defer wg.Done()
-		err := client.AddDatapoints(childCtx, points)
-		if err != nil {
-			span.Error(err)
-			span.Add(ssf.Count("flush.error_total", 1, map[string]string{"cause": "io", "sink": "signalfx"}))
-			errorCh <- err
+		var firstErr error
+		perFlush := maxPerFlush
+		if perFlush == 0 {
+			perFlush = len(points)
+		}
+		errs := make(chan error)
+		flushes := 0
+		for i := 0; i < len(points); i += perFlush {
+			end := i + perFlush
+			if end > len(points) {
+				end = len(points)
+			}
+			go submitDatapoints(childCtx, cl, client, points[i:end], errs)
+			flushes++
+		}
+		for i := 0; i < flushes; i++ {
+			err := <-errs
```

Comment:
Is it possible for any of these requests to time out (and therefore block forever here)?

Reply:
(we set a deadline on the contexts)